### ISSUE TITLE
Sprig script is not getting loaded when template caching is enabled

### ISSUE THREAD

**Issue by WHITE-developer** [putyourlightson/craft-sprig#329]
Please provide a clear and concise description of what the bug is, including:
- Steps to reproduce the issue.
  Have a sprig component in a cached template
  On the first request the htmx script is loading
  On a next request the script isn't added anymore.
- A clear and concise description of what you expected to happen.
  htmx script should get loaded always
- If applicable, add screenshots to help explain your problem.
- The plugin and Craft version numbers.
craft cms 4.5.6.1 and sprig 2.7.1

**Comment by bencroker** [comment]
How exactly is the template being cached?

**Comment by WHITE-developer** [comment]
Matrix blocks are getting cached by default cache tags:
```
{% for block in blocks %}
    {% cache using key "block.type.#{block.type.handle}.id.#{block.id}" %}
        {{ include('_blocks/' ~ block.type, ignore_missing=true) }}
    {% endcache %}
{% endfor %}
 ```
 
 And then inside a block template a sprig component is getting called `{{ sprig('_components/buy-btn', { _variantId: product.defaultVariantId }) }}`
 
 When setting the .env var `CRAFT_ENABLE_TEMPLATE_CACHING` to `true` the first request is working, on the next request htmx (assetbundle) isn't loading in, until I clear the template cache or set the .env variable to `false`

**Comment by bencroker** [comment]
Thanks for clarifying. I just added the `sprig.registerScript()` in version 2.7.2 that you can add to a layout template to ensure that the script is always registered.

```twig
{% do sprig.registerScript() %}
```

**Comment by MoritzLost** [comment]
@bencroker Just stumbled across this as well. This was pretty unexpected, since Craft has been able to keep track of JS/CSS assets that have been added through the `{% js %}` tag for a long time now. I think I figured out why the script is not loading. Craft keeps track of all JS/CSS assets registered through the `View` class and saves them alongside the cached HTML, then adds them again when the cache is used:

https://github.com/craftcms/cms/blob/642966286e13e9edf40180d6dbe50c64c1a150bf/src/services/TemplateCaches.php#L120-L129

https://github.com/craftcms/cms/blob/642966286e13e9edf40180d6dbe50c64c1a150bf/src/services/TemplateCaches.php#L85-L96

But Sprig is registering the HTMX script as an _Asset Bundle_ instead, which is not kept track of inside the `{% cache %}` tag! Probably because those are more commonly used in the CP?

In any case, registering the script globally is a bit awkward. Either we load it on _every_ page (even though it may not be used on a lot of pages), or put complex logic in a layout file to determine whether the current page contains a Sprig component.

Is there any way to fix this issue for good to avoid this gotcha? This could be solved in Core by keeping track of registered Asset Bundles, but that sounds challenging. Maybe Sprig can register the script in a way that will be picked up by the template cache service?

**Comment by bencroker** [comment]
Thanks for the explanation, @MoritzLost. Ironically, Sprig used to register the htmx script using `View::registerJsFile()`, but this behaviour was changed to `View::registerAssetBundle` in https://github.com/putyourlightson/craft-sprig-core/pull/8 to add support for Craft Cloud. 

I’ve attempted a [workaround here](https://github.com/putyourlightson/craft-sprig-core/commit/594aec03857cbbad92b794787dcfdd90d1f29752#diff-520280e5d19f4351dd44b0d26bb8b1d14f23bbf5d94571e00c26f50b52996f0e) that registers the JS file after registering the asset bundles, which works in my tests. I’ll just have to pass this by the P&T team to ensure this continues to work with Cloud.

**Comment by MoritzLost** [comment]
@bencroker Ah, interesting to know the story behind that change. Yeah, sounds like something that should be addressed by P&T, especially if this is the preferred way to include assets in Craft Cloud. Maybe Asset Bundles should really be made compatible with template caching, though this might be complicated to implement.

**Comment by bencroker** [comment]
P&T have let me know that they'll discuss this internally. I'll update here when I know more.

**Comment by bencroker** [comment]
It appears this may be rolled into core in a future version. Until then, the workaround will have to do and I’ll be cutting a release this week.

**Comment by MoritzLost** [comment]
@bencroker That's great, thanks Ben!

**Comment by bencroker** [comment]
Released in versions [2.9.0](https://github.com/putyourlightson/craft-sprig/releases/tag/2.9.0) and [3.1.0](https://github.com/putyourlightson/craft-sprig/releases/tag/3.1.0).
